btree/btree.c

```diff
@@ -1501,6 +1502,253 @@ static void generic_move(struct nd *src, struct nd *tgt,
+ * to @accum.
+ */
+static void m0_btree_free_node_credit(const struct m0_btree  *tree,
+				      m0_bcount_t             nr,
+				      struct m0_be_tx_credit *accum)
+{
+	struct m0_be_tx_credit cred = {};
+
+	btree_mem_free_credit(tree, max_node_size, &cred);
+	m0_be_tx_credit_mac(accum, &cred, nr);
+}
+
+static void btree_delete_key_value_credit(const struct m0_btree  *tree,
+					  m0_bcount_t             ksize,
+					  m0_bcount_t             vsize,
+					  struct m0_be_tx_credit *accum)
+{
+	btree_mem_free_credit(tree, ksize + vsize, accum);
+	m0_be_tx_credit_add(accum,
+			    &M0_BE_TX_CREDIT_TYPE(struct m0_btree_rec));
+}
+
+/**
+ * Calculates the credit needed to update particular record ,
+ * which requires deleting old record and inserting new record with modified key/value.
+ * As we might want to update delimiter key in delete operation, credits for record updates are needed
+ */
+static void btree_update_delimeter_key_credit(const struct m0_btree  *tree,
+					      struct m0_be_tx_credit *accum)
+{
+	btree_delete_key_value_credit(tree, max_ksize, internal_vsize, accum);
+	btree_insert_key_value_credit(tree, max_ksize, internal_vsize, accum);
+}
+
+static void btree_delete_leaf_node_credit(const struct m0_btree  *tree,
+					  m0_bcount_t             ksize, 
+					  m0_bcount_t             vsize, 
+					  struct m0_be_tx_credit *accum)
+{
+	btree_delete_key_value_credit(tree, ksize, vsize, accum);
+	btree_move_key_value_credit(accum, MIN_MOVE);
+	m0_btree_free_node_credit(tree, 1, accum);
+	btree_update_delimeter_key_credit(tree, accum); //credits to update delimeter key
+}
+
+static void btree_delete_internal_node_credit(const struct m0_btree  *tree,
+					      struct m0_be_tx_credit *accum)
+{
+	btree_delete_key_value_credit(tree, max_ksize, internal_vsize, accum);
+	btree_move_key_value_credit(accum, MIN_MOVE);
+	m0_btree_free_node_credit(tree, 1, accum);
+	btree_update_delimeter_key_credit(tree, accum);
+}
+
+static void delete_root_credit(const struct m0_btree  *tree,
+			       struct m0_be_tx_credit *accum)
+{
+	btree_delete_key_value_credit(tree, max_ksize, internal_vsize, accum);
+	btree_move_key_value_credit(accum, ALL_MOVE);
+	m0_btree_free_node_credit(tree, 1, accum);
+}
+
+/**
+ * Calculates total credits required to perform @nr delete operations
+ * @param nr     Number of delete operations.
+ * @param ksize  Key data size.
+ * @param vsize  Value data size.
+*/
+static void delete_credit(const struct m0_btree  *tree,
+			  m0_bcount_t             nr,
+			  m0_bcount_t             ksize,
+			  m0_bcount_t             vsize,
+			  struct m0_be_tx_credit *accum)
+{
+	uint32_t height = BTREE_HEIGHT_MAX;
+
+	btree_delete_leaf_node_credit(tree, ksize, vsize, accum);
+
+	struct m0_be_tx_credit cred = {};
+	btree_delete_internal_node_credit(tree, &cred);
+	m0_be_tx_credit_mac(accum, &cred, height - 2);
+	
+	delete_root_credit(tree, accum);
+
+	m0_be_tx_credit_mul(accum, nr);
```

Comment:
Similar issues with "insert_credit()"

Reply:
Explained in insertion part
[link](https://github.com/Seagate/cortx-motr/pull/594#discussion_r629852782)

Reply:
Also, we need to have a local variable to estimate the credit, and then multiply it with "nr":
```
	uint32_t height = BTREE_HEIGHT_MAX;
	struct m0_be_tx_credit cred = {};
	struct m0_be_tx_credit local_accum = {};


	btree_delete_leaf_node_credit(tree, ksize, vsize, &cred);

	btree_delete_internal_node_credit(tree, &local_accum);
	m0_be_tx_credit_mac(&local_accum, &cred, height - 2);

	delete_root_credit(tree, &local_accum);

	m0_be_tx_credit_mul(&local_accum, nr);


	m0_be_tx_credit_add(accum, &local_accum);

```

Reply:
Yes I got it I will  make changes